server/src/main/kotlin/org/jetbrains/bsp/bazel/server/common/ServerContainer.kt

```diff
@@ -79,9 +75,11 @@ class ServerContainer internal constructor(
       val bazelProjectMapper =
         BazelProjectMapper(languagePluginsService, bazelPathsResolver, targetKindResolver, bazelInfo, bspClientLogger, metricsLogger)
       val targetInfoReader = TargetInfoReader()
+      val currentContext = workspaceContextProvider.currentWorkspaceContext()
+      val bazelExternalRulesQuery = if (currentContext.enabledRules.values.isEmpty()) BazelExternalRulesQueryImpl(bazelRunner, bazelInfo.isBzlModEnabled) else BazelEnabledRulesQueryImpl(currentContext.enabledRules)
```

Comment:
please put the implementation part inside `BazelExternalRulesQueryImpl#fetchExternalRules`.
You can pass additional argument `enabledRules` to `BazelExternalRulesQueryImpl` also.

Reply:
Done!